doc/crdt-details.md

```diff
@@ -0,0 +1,1086 @@
+
+The fact that, even in a merge, `Revision`s are only ever appended leads to the interesting fact that two peers (Separate devices/engines that share state my merging) can have `Engine`s that represent the same document contents and history, but where the `Revision`s are in a totally different order. This is fine though because the `Revision` ids allow us to compare the identity of two `Revision`s even if their `Subset`s are different due to transforms, and undo groups allow us to maintain and manipulate undo history order separately from CRDT history order.
+
+In practice the order of the `Revision` history will tend to be very similar between peers. Any edit that occurs while another edit is visible on screen (present in this peer's `Engine`) will never be re-ordered before that visible edit. Thus if there are no concurrent edits made on devices that sync with each other by merging, the devices will end up with the same `Revision` history.
+
+**Note:** As of the time this was written, Xi's `merge` implementation does not support undo operations, and will panic if one is encountered. We plan to fix this and expect that the structure of the merge operation will stay the same, just with more cases. So for the rest of this description, presume `Edit`s are the only type of `Revision`.
+
+#### Base and common revisions
+
+An important part of merging is figuring out which revisions the two sides have in common. These common revisions are found as two parts:
+
+1. The "base": A prefix of the same length of both histories such that the set of revisions in both prefixes is the same.
+    - Note that they aren't necessarily in the same order, just all shared. That's because all indices are relative to the union string, which is the same regardless of the order of the revisions that created it.
+    - We ignore everything in this prefix, so the longer it is the faster the merge runs, but a length of 0 is perfectly fine and in fact at the time this was written, that's what Xi uses. However, we plan on finding the maximal such prefix as an important optimization.
+1. The "common" revisions: After the base, some revisions on each side will be shared by both sides. The common revisions are the intersection of the two revision sets after the base.
+    - Note that the common revisions aren't necessarily in the same positions or order on each side.
+
+![CRDT Merge Flow Example](img/merge-intro.png)
+
+If you're wondering how you can end up with common revisions not in the base, and in different positions and ordering, that's because it's rare. But these cases can occur even with only two peers. See the example below.
```

Comment:
strictly speaking, it may not be the case that the reason I'm wondering how this is the case is "because it's rare". You could say "well: it's rare." (or something else conversational) or you could just jump to the explanation.